Refactor TritonTemplate.generate and move codgen part to generate_and_load

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #151764

Splitting https://github.com/pytorch/pytorch/pull/149267/ .
This first PR just refactor the code without adding any caching functionality.
The logic of generating the code and loading it is moved to generate_and_load() + some typing

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

laithsakka committed:
refactor

[ghstack-poisoned]

laithsakka commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1078,52 +1079,40 @@ def __init__(self, name: str, grid: Any, source: str, debug=False) -> None:
-            mutated_inputs: Optional list of input nodes that are mutated by the kernel, this is helpful
-                if you need to return multiple outputs. You can pass them as inputs and mark them as
-                being mutated by the kernel.
-        """
```
this code is moved to generate_and_load as is

laithsakka commented on torch/_inductor/select_algorithm.py:
```diff
@@ -946,8 +947,27 @@ def hook():
         return "<STORE_OUTPUT>"
 
     def render(self, template, kwargs):
+        def make_template_env():
```
this is only used in this function so i made it local, in next PR i also want it local

drisspg replied:
nit: I am okay with this but I think a nontrivially hard  things about grokking pt2 stack is the insane usage local funcs, following the execution flow for any given functions is wild.


So if weant this local can we just make it variable and not a func?

laithsakka replied:
ah forgot to address rthis

laithsakka replied:
done

laithsakka committed:
Update on "Refactor TritonTemplate.generate and move codgen part to generate_and_load"


Splitting https://github.com/pytorch/pytorch/pull/149267/ this first PR just refactor the code with out adding any 
caching functionality.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1078,52 +1079,40 @@ def __init__(self, name: str, grid: Any, source: str, debug=False) -> None:
+            dict[str, Any],
+        ]
+    ]:
+        # breakpoint()
```
remove this line

drisspg commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1078,52 +1079,40 @@ def __init__(self, name: str, grid: Any, source: str, debug=False) -> None:
+        layout: ir.Layout,
+        kwargs: dict[str, Any],
+    ) -> Optional[
+        tuple[
```
Nit: an Type defined might be easier for this bad boy return

laithsakka replied:
updated to use NamedTuple.

drisspg commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1162,17 +1149,37 @@ def generate(  # type: ignore[override]
+            )
+
+        def generate_code(kernel) -> Optional[tuple[str, str]]:
+            def make_extra() -> str:
```
similar comment to above but non blocking

laithsakka committed:
Update on "Refactor TritonTemplate.generate and move codgen part to generate_and_load"


Splitting https://github.com/pytorch/pytorch/pull/149267/ .
This first PR just refactor the code without adding any caching functionality.
The logic of generating the code and loading it is moved to generate_and_load() + some typing

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
@pytorchbot merge

laithsakka committed:
Update on "Refactor TritonTemplate.generate and move codgen part to generate_and_load"


Splitting https://github.com/pytorch/pytorch/pull/149267/ .
This first PR just refactor the code without adding any caching functionality.
The logic of generating the code and loading it is moved to generate_and_load() + some typing

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Refactor TritonTemplate.generate and move codgen part to generate_and_load"


Splitting https://github.com/pytorch/pytorch/pull/149267/ .
This first PR just refactor the code without adding any caching functionality.
The logic of generating the code and loading it is moved to generate_and_load() + some typing

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Refactor TritonTemplate.generate and move codgen part to generate_and_load"


Splitting https://github.com/pytorch/pytorch/pull/149267/ .
This first PR just refactor the code without adding any caching functionality.
The logic of generating the code and loading it is moved to generate_and_load() + some typing

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
@pytorchbot  merge

laithsakka commented:
@pytorchbot merge
